### PR TITLE
nfa: add the NFA abstraction

### PR DESCRIPTION
**Pull request by TimaFrolov**

### CODE REVIEW

File: lib/nfa.mli [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+
+type ('state, 'varname) dfaCollisions =
+  ('state * ('varname, bit) Map.t * ('varname, bit) Map.t * 'state * 'state)
+  list
+
+val create_dfa :
+     transitions:('s * ('var, bit) Map.t * 's) list
+  -> start:'s
+  -> final:'s list
+  -> ('var dfa, ('s, 'var) dfaCollisions) result
```
Comment (georgiy-belyanin) [inline lib/nfa.mli:36]: Probably it's worth it using GADT here too? I think it's better to hide the states the same way as it's done in NFA.

File: test/unit/TestNfa.ml [hunk 2]
```diff
@@ -0,0 +1,162 @@
+open Lib
+open Bits
+module Map = Nfa.Map
+
+let ( let* ) = Option.bind
+
```
Comment (georgiy-belyanin) [inline test/unit/TestNfa.ml:6]: It should be in the utils.

**Review by georgiy-belyanin — DISMISSED** [review]
I've made a very quick review though. Probably there's a lot of space to write less though the code seems to work rather good.

Please fix the formatter and apply the comments if you want to.

File: lib/nfa.mli (hunk 1 above)
Comment (TimaFrolov) [reply lib/nfa.mli:36]: The DFA itself hides states the same way NFA does, but I think states shluld be a part of error to provide better error description

File: test/unit/TestNfa.ml (hunk 2 above)
Comment (TimaFrolov) [reply test/unit/TestNfa.ml:6]: Should it really? 🤔
`let*` could be used as symlink to bind function for any monad, not just Option. 
If you think it's unlikely we will use other monads in the future, I can move it to Utils

Comment (georgiy-belyanin) [reply test/unit/TestNfa.ml:6]: Ah, missed this point. I'm sorry.

**Review by georgiy-belyanin — APPROVED** [review]
LGTM. Thank you.

**Pull request merged**